smack-extensions/src/main/java/org/jivesoftware/smackx/softwareinfo/SoftwareInfoManager.java

```diff
@@ -0,0 +1,144 @@
+ */
+package org.jivesoftware.smackx.softwareinfo;
+
+import java.io.IOException;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+import org.jivesoftware.smack.Manager;
+import org.jivesoftware.smack.SmackException.NoResponseException;
+import org.jivesoftware.smack.SmackException.NotConnectedException;
+import org.jivesoftware.smack.XMPPConnection;
+import org.jivesoftware.smack.XMPPException.XMPPErrorException;
+import org.jivesoftware.smack.parsing.SmackParsingException;
+import org.jivesoftware.smack.xml.XmlPullParserException;
+import org.jivesoftware.smackx.disco.ServiceDiscoveryManager;
+import org.jivesoftware.smackx.disco.packet.DiscoverInfo;
+import org.jivesoftware.smackx.disco.packet.DiscoverInfo.Identity;
+import org.jivesoftware.smackx.disco.packet.DiscoverInfoBuilder;
+import org.jivesoftware.smackx.softwareinfo.element.SoftwareInfoElement;
+
+import org.jxmpp.jid.Jid;
+
+/**
+* Entry point for Smack's API for XEP-0232: Software Information.
+* <br>
+* To add {@link SoftwareInfoElement}, use {@link #addSoftwareInfoElement(SoftwareInfoElement)} method.
+* <br>
+* In order to be updated with other user's Software Information add {@link SoftwareInfoListener}, use {@link #addSoftwareInfoListener(SoftwareInfoListener)}.
+* <br>
+* @see <a href="https://xmpp.org/extensions/xep-0232.html">
+*     XEP-0232 : Software Information.
+*     </a>
+*/
+public final class SoftwareInfoManager extends Manager {
+
+    private static final String FEATURE = "http://jabber.org/protocol/disco";
+    private static final Map<XMPPConnection, SoftwareInfoManager> INSTANCES = new WeakHashMap<>();
+    private final Set<SoftwareInfoListener> softwareInfoListeners = new HashSet<>();
+    private final ServiceDiscoveryManager serviceDiscoveryManager;
+    private DiscoverInfoBuilder discoverInfoBuilder;
+
+    public static synchronized SoftwareInfoManager getInstanceFor (XMPPConnection connection) throws IOException, XmlPullParserException, SmackParsingException {
+        SoftwareInfoManager manager = INSTANCES.get(connection);
+        if (manager == null) {
+            manager = new SoftwareInfoManager(connection);
+            INSTANCES.put(connection, manager);
+        }
+        return manager;
+    }
+
+    private SoftwareInfoManager(XMPPConnection connection) throws IOException, XmlPullParserException, SmackParsingException {
+        super(connection);
+        serviceDiscoveryManager = ServiceDiscoveryManager.getInstanceFor(connection);
+        addFeature();
+        discoverInfoBuilder = DiscoverInfo.builder(connection);
+    }
+
+    /**
+     * Refer to <a href="https://xmpp.org/registrar/disco-categories.html">XMPP Registry for Service Discovery Identities</a>.
+     * @param category Category of the Identity.
+     * @param name Name of the Identity.
+     * @param type Type of the Identity.
+     */
+    public void setIdentity(String category, String name, String type) {
```

Comment:
I'm not sure if this is the correct approach, as I as a user would expect the identity to be published once I set it with this method. That however is not the case.

I'd recommend to create a wrapper around the DiscoInfoBuilder that exposes methods related to the context of XEP-0323 and make the publishing one atomic call.

Reply:
@vanitasvitae I have reflected upon this, feel free to take a look when you have time :)